### PR TITLE
fix(mcp): constrain spec codegen paths

### LINKED ISSUE #3358: Constrain MCP spec codegen input/output paths to the repository workspace
# Constrain MCP spec codegen input/output paths to the repository workspace

## Summary

The `ae_spec_codegen` MCP tool accepts caller-supplied `irPath` and `outDir` values, resolves them, and forwards them to the codegen CLI. The CLI and deterministic generator then read the selected AE-IR file and write generated files and a manifest under the resolved output directory without verifying that the paths remain under an approved workspace or generated-artifact directory. Some generated file paths also include IR-derived entity names without treating them as a single safe path component.

## Security impact

A compromised or prompt-injected MCP client with access to this tool can cause generated artifacts to be written outside the intended repository workspace, subject to the filesystem permissions of the MCP server process. This can overwrite or create local files in adjacent writable locations and can also place generated code where later tooling may consume it. The audited checkout did not include the built `dist/src/cli/index.js`, so runtime reachability depends on the built package/deployment context.

## Evidence

- `repo/src/mcp-server/spec-synthesis-server.ts:63-67` accepts `irPath` and `outDir` as strings.
- `repo/src/mcp-server/spec-synthesis-server.ts:183-191` resolves the caller paths and invokes the codegen CLI child process with those values.
- `repo/src/cli/codegen-cli.ts:73-82` resolves `--input` and `--output` into generator options without workspace containment.
- `repo/src/codegen/deterministic-generator.ts:221-226` writes generated files to `join(this.options.outputDir, file.filePath)`.
- `repo/src/codegen/deterministic-generator.ts:621-623` writes the manifest under the same output directory.
- `repo/src/codegen/deterministic-generator.ts:276-347` includes IR entity names in generated React/database file paths before the final write.

## Recommended remediation

Require all MCP file reads/writes to pass through a shared path policy: resolve and normalize against a configured workspace root, reject absolute paths and `..` escapes unless an explicit allowlist permits them, and verify the final write path remains under the approved output directory. Treat IR-derived names as identifiers, not path strings: validate entity/function names against language-appropriate identifier rules and separately derive collision-safe filenames. Fail closed on invalid paths before spawning the child process.

## Regression test idea

Add unit tests for `ae_spec_codegen`/codegen path policy that cover valid workspace-relative paths, absolute paths, dot-segment paths, and IR entity names containing path separators. Assert rejected paths never reach `spawnSync` or `writeFileSync`, and valid generation writes only under the configured generated-artifact directory.

### LINKED ISSUE #3360: Constrain ae_spec_compile output paths to an approved workspace
# Constrain ae_spec_compile output paths to an approved workspace

## Summary

The `ae_spec_compile` MCP tool accepts caller-supplied `inputPath` and optional `outputPath` values. The handler resolves those strings and passes them to `AESpecCompiler.compile()`, which writes the compiled AE-IR JSON directly to `outputPath` when supplied. The path is not checked against an approved repository workspace or generated-artifact directory.

## Security impact

A compromised or prompt-injected MCP client with access to `ae_spec_compile` can create or overwrite a JSON file path reachable by the MCP server process, subject to filesystem permissions and successful compilation. This can move generated AE-IR outside the intended workspace or clobber adjacent writable files. The reviewed compile path writes data; it does not execute the written file.

## Evidence

- `repo/src/mcp-server/spec-synthesis-server.ts:50-55` accepts `inputPath` and `outputPath` as strings.
- `repo/src/mcp-server/spec-synthesis-server.ts:119-130` resolves those values and forwards `outputPath: resolve(parsed.outputPath)` to `AESpecCompiler.compile()`.
- `repo/packages/spec-compiler/src/compiler.ts:13-18` reads the selected input file.
- `repo/packages/spec-compiler/src/compiler.ts:31-33` writes the compiled IR JSON directly to `outputPath` with `writeFileSync()`.
- A safe local path-normalization model showed representative `../` and absolute output paths resolve outside the repository boundary; no file was written during validation.

## Recommended remediation

Apply a shared MCP filesystem policy to `ae_spec_compile`: resolve relative paths against a configured workspace or artifact root, reject absolute paths and dot-segment escapes unless explicitly allowlisted, and verify the normalized final output path remains under the approved directory before any write. Prefer a fixed generated-artifact directory for compile outputs.

## Regression test idea

Add tests for valid workspace-relative paths, `../` traversal, absolute paths, and normalization edge cases. Mock `writeFileSync()` and assert rejected paths never reach the write sink, while valid outputs are written only under the configured artifact root.

### PR DESCRIPTION
**Pull request by ootakazuhiko**
## Summary
- Add a shared workspace path policy for MCP/CLI codegen paths, rejecting absolute paths, dot-segment escapes, backslash separators, and symlink ancestors that resolve outside the approved workspace.
- Apply the policy to `ae_spec_compile`, `ae_spec_validate`, and `ae_spec_codegen` before compiler writes or codegen child-process spawning.
- Apply the same workspace-relative policy to codegen CLI generate/watch inputs and outputs.
- Contain deterministic generator file writes under the configured output directory and validate IR-derived identifiers before using them in generated filenames/code.

Closes #3358
Closes #3360

## Acceptance
- MCP `ae_spec_compile` output paths are resolved only under the approved MCP workspace before `AESpecCompiler.compile()` can write.
- MCP `ae_spec_codegen` `irPath`/`outDir` are resolved only under the approved workspace, and child-process arguments remain workspace-relative.
- Codegen CLI rejects absolute or dot-segment input/output paths instead of resolving them into arbitrary filesystem locations.
- Generated files and manifests are contained under the configured output directory, and unsafe IR-derived entity/field/API method identifiers fail closed.
- Regression tests cover valid paths, traversal/absolute rejection, symlink ancestor rejection, child-process argument shaping, and IR-derived path traversal prevention.

## Verification
- `pnpm -s exec vitest run tests/unit/utils/workspace-path-policy.test.ts tests/unit/mcp-server/spec-synthesis-path-policy.test.ts tests/unit/codegen/deterministic-generator-path-policy.test.ts --reporter dot`
- `pnpm -s run types:check`
- `pnpm -s exec eslint --quiet src/utils/workspace-path-policy.ts src/mcp-server/spec-synthesis-path-policy.ts src/mcp-server/spec-synthesis-server.ts src/cli/codegen-cli.ts src/codegen/deterministic-generator.ts tests/unit/utils/workspace-path-policy.test.ts tests/unit/mcp-server/spec-synthesis-path-policy.test.ts tests/unit/codegen/deterministic-generator-path-policy.test.ts`
- `pnpm -s run build`
- `pnpm -s run check:schemas`
- `pnpm -s run check:doc-consistency`
- `git diff --check`
- CLI smoke: relative codegen generate succeeds; absolute input path is rejected with the workspace-relative policy error.

## Rollback
- Revert this PR to restore the previous MCP/CLI codegen path handling. If reverted, do not expose `ae_spec_compile` or `ae_spec_codegen` to untrusted MCP clients until equivalent workspace containment is restored.

### CODE REVIEW

File: src/mcp-server/spec-synthesis-server.ts [hunk 1]
```diff
@@ -182,24 +188,23 @@ async function start() {
...
-          const irPath = resolve(parsed.irPath);
+          const paths = resolveSpecCodegenPaths(parsed.irPath, parsed.targets, parsed.outDir);
           type IRLike = { domain?: unknown[]; api?: unknown[] };
-          const ir = JSON.parse(readFileSync(irPath, 'utf-8')) as unknown as IRLike;
+          const ir = JSON.parse(readFileSync(paths.irPath, 'utf-8')) as unknown as IRLike;
           const { spawnSync } = await import('child_process');
-          const outBase = parsed.outDir || 'generated';
           const run = (t: string, dir: string) =>
-            spawnSync(process.execPath, ['dist/src/cli/index.js', 'codegen', 'generate', '-i', irPath, '-o', resolve(dir), '-t', t], { stdio: 'inherit' });
+            spawnSync(process.execPath, ['dist/src/cli/index.js', 'codegen', 'generate', '-i', paths.irPathArg, '-o', dir, '-t', t], { stdio: 'inherit' });
```
[the remark this answers is not in the document]
Comment (ootakazuhiko) [reply src/mcp-server/spec-synthesis-server.ts]: 対応しました。`ae_spec_codegen` の spawn で `cwd` を MCP workspace に固定し、`AE_CODEGEN_WORKSPACE_ROOT` も同じ workspace に渡すようにしました。これにより、親が検証した workspace-relative `irPath` / `outDir` を子 CLI が同一ルートで解釈します。併せて `spawnSync` の `error` / `status` を確認し、子プロセス失敗時に成功扱いで結果を返さないようにしています。回帰として child process options の unit test も追加しました。
